### ISSUE TITLE
config.ini?

### ISSUE THREAD

**Issue by dpmusic** [janvv/cgm-dashboard#1]
I don't understand where the config.ini file comes from- do I somehow generate that on my own? Sorry, newb, but I would dearly love to get this running for my daughter....

**Comment by StephenBrown2** [comment]
The details are given in the README: https://github.com/janvv/cgm-dashboard#installation
Looks like depending on how you want to connect to your data, you give it a different set of config. E.g. for connecting to your Nightscout instance, put this with your details in `config.ini`:

```ini
[REST]
domain = <the domain of your REST service>
port = <the port of your REST service>
```

If you're using token auth, you'll need to prepend that to the domain, since all it's doing is combining them to make the url: `https://{domain}:{port}/api/v1/entries/sgv.json`

**Comment by janvv** [comment]
Thanks @StephenBrown2 for stepping in. Yes @dpmusic, you will have to configure your personal config file that specifies the details to your nightscout mongo server. For REST, i currently do not support username and passwords. I implemented  REST for a friend that uses a ns.10be.de server which does not support authentication.

**Comment by dpmusic** [comment]
Amazing of you guys to respond, thank you! I SO appreciate your support with this.

Ok, I think I’m getting closer, but struggling with a couple of things-

I have nightscout set up on Heroku, so I understand I should be good to go with the MongoDB sign in. However I can’t quite understand how it correlates:

Here is the MONGODB_URI given in my Heroku config table (MONGO_COLLECTION is set to “entries)": 

"mongodb://heroku_jh0tj7tr:mj52guoa2pj1mbtvmhotqa102g@ds145574.mlab.com:45574/heroku_jh0tj7tr"

So far I think I’ve been able to tell from this:

user = /heroku_jh0tj7tr
password = mj52guoa2pj1mbtvmhotqa102g
host = (this is the one I’m less sure about) <something>.mlab.com?
database = heroku_jh0tj7tr
port = 45574

Am I at all close?

**Comment by StephenBrown2** [comment]
You are quite close. I plugged the URL after substituting some values into https://www.freeformatter.com/url-parser-query-string-splitter.html and you can do the same to see for yourself. Here are the key elements:

URL Parts for `mongodb://heroku_user:thisismypassword@ds145574.mlab.com:45574/heroku_database`
Username: 	heroku_user
Password: 	thisismypassword
Hostname: 	ds145574.mlab.com
Port: 	45574
File name (Database): 	heroku_database

In mlab's case, they make the database the same as the username, so although in this case there is `heroku_user` and `heroku_database`, in yours they would both be `heroku_jh0tj7tr`.

```
user = heroku_jh0tj7tr
password = you_should_probably_change_your_password_now
host = ds145574.mlab.com
database = heroku_jh0tj7tr
port = 45574
```

**Comment by dpmusic** [comment]
Extremely close, I think! But I have run into another issue that I can’t solve after another few hours on google. So I must ask for help again-

I did:

[MongoDB]
user = heroku_jh0tj7tr
password = mj52guoa2pj1mbtvmhotqa102g@ds145574
host = ds145574.mlab.com
database = heroku_jh0tj7tr
port = 45574

But got this error:
pymongo.errors.InvalidURI: Username and password must be escaped according to RFC 3986, use urllib.parse.quote_plus().

So I tried this:

[MongoDB]
user = urllib.parse.quote_plus('heroku_jh0tj7tr')
password = urllib.parse.quote_plus('mj52guoa2pj1mbtvmhotqa102g@ds145574')
host = ds145574.mlab.com
database = heroku_jh0tj7tr
port = 45574

But now I get this error:
line 61, in parse_userinfo
    "RFC 3986, use %s()." % quote_fn)

No combination of single quotes or double quotes seems to help. I’m in over my head. Many thanks again.

**Comment by StephenBrown2** [comment]
You've got two things going on now, first, your password has gotten the first part of the host attached to it, and in the second example, you're trying to use python in an ini file, which really probably won't work.

```ini
[MongoDB]
user = heroku_jh0tj7tr
password = mj52guoa2pj1mbtvmhotqa102g
host = ds145574.mlab.com
port = 45574
database = heroku_jh0tj7tr
```
Note that there is no `@ds145574` at the end of the password.
This code in the app: https://github.com/janvv/cgm-dashboard/blob/master/adapter.py#L19
will turn that into:
```
'mongodb://heroku_jh0tj7tr:mj52guoa2pj1mbtvmhotqa102g@ds145574.mlab.com:45574/heroku_jh0tj7tr'
```
which should match your given connection URI:
```
"mongodb://heroku_jh0tj7tr:mj52guoa2pj1mbtvmhotqa102g@ds145574.mlab.com:45574/heroku_jh0tj7tr"
```

**Comment by dpmusic** [comment]
Ok, I see, yes, a stupid mistake there which led me down a rabbit hole. Fixed that, and…..

It’s working!! Thank you very much for your patience.

**Comment by dpmusic** [comment]
Will this nightscout/mlab/mongodb change have any impact on the function of cgm-dashboard?

**Comment by StephenBrown2** [comment]
Yes, once you migrate from mLab, your database connection information will change. So you may want to change your config to use your nightscout url and credentials.

**Comment by dpmusic** [comment]
Wow, my timing was just terrible there. 

Ok well, looks like I must ask for more help, as I’ve once again exhausted google.

I take it the domain is not “myAPISecret”@https://darcy-phillips-ns.herokuapp.com <https://darcy-phillips-ns.herokuapp.com/>, as I think I remember you saying you didn’t support usernames and passwords. Would my domain just be https://darcy-phillips-ns.herokuapp.com <https://darcy-phillips-ns.herokuapp.com/>, then?

Also, I can’t determine what the port number would be, or how to find out. 

I’ve tried various combinations, with and without API secret, and with port = 80, port = 443, port = 5000. Totally guessing, have no idea. (I don’t believe I’m using token auth, doesn’t sound familiar).

my config.ini looks like this:

[REST]
domain = https://darcy-phillips-ns.herokuapp.com
port = 80

and this is what happens in my session:



Thank you again.
